### ISSUE TITLE
[Ingest Plugin] continue_on_failure parameter

### ISSUE THREAD

**Issue by talevy** [elastic/elasticsearch#14548]
It would be nice to have a flag for ingest pipelines that run during index/bulk requests to allow for some configuration of how errors are handled during ingestion.

Currently, any exception that is caused during an ingest pipeline, the pipeline halts and the document being ingested is not indexed. Some environments may prefer to not block on such failures and continue to index the failed document anyways. For these scenarios it would be nice to have an `continue_on_failure` flag turned on to instruct Elasticsearch to index the document as it was before it was processed by the ingest pipeline. Such documents can be tagged so that they can be found in the index for re-processing.

`_simulate` should also support this flag to help with recreating pipeline behavior

depends on https://github.com/elastic/elasticsearch/pull/14888

**Comment by martijnvg** [comment]
@talevy maybe we should name the param `continue_on_failure`?

**Comment by talevy** [comment]
deal! I'll update

**Comment by javanna** [comment]
I think this flag is needed to decide whether to go on with following processors in the pipeline and still index the current document. It nicely applies to the index api, and the bulk api in the context of each single document.

What currently happens when something fails while indexing using bulk though, is that we fail the whole request, which doesn't seem convenient. I think ideally we should be able to go ahead with the other documents and add the failure to the response for that specific document even when `continue_on_failure` is set to `false`. This might be a high hanging fruit though and might require a separate issue/discussion, as `continue_on_failure` requires some more thinking on how to apply it to the bulk api.

**Comment by martijnvg** [comment]
I think there is no doubt about whether we need this option.

I agree that in the bulk api providing item level granularity control for this flag is tricky, because we we can't set the failure response item in the bulk response. We can only decide whether to delegate to the bulk api. To achieve this I think we also need to have logic in `IngestActionFilter#apply(String, ActionResponse, ActionListener, ActionFilterChain)` (which currently just delegates).

**Comment by javanna** [comment]
We sorted out the bulk api aspect with #14888, meaning that whenever a processor fails for a document it won't affect the other documents within the same bulk request.

In this issue we want focus on the behaviour for each single document. Let's say one processor fails as part of the pipeline execution, we want to be able to control what happens with that single document. The main questions are: should we go ahead with the following processors in the pipeline and index the document? If so how do we keep track of the fact that something failed?

**halt on failure**: if we don't continue, we return an error in the response (either the index api fails or that item will be marked as failed as part of the bulk response) and the client knows what went wrong. The document doesn't get indexed. I can't think of anything more that could be done in this case. This is how things currently work already.

**continue on failure**: if we continue, the client will not be notified of the error, hence we have to at least tag the document so that we keep track of what went wrong. We may also want to index the document somewhere else (different index). This requires additional configuration (which index and which field keeps track of what failed) hence we need to decide how to make it configurable.

**the error**:  easier things first, it seems clear that we want to be able to add a structured error to the document in a certain field (potentially configurable?), actually an array of errors given that more processors can fail at the same time if we continue on failure. Each item would be an object containing the pipeline id, the processor id and structured information about the error, similar to the structured errors that elasticsearch now returns from the rest layer.

Should the possible actions be limited to the ones described (pick the index, eventually pick the error field), or should we allow to configure a specific pipeline to be executed on failure so that whatever can be done through processors can be done on failure as well? The latter seems very flexible, maybe too flexible? Also potentially dangerous in case that pipeline ends up failing (on failure). The former is simpler and might be limiting but I think it would be good to try and come up with more options that we would need before going for the whole pipeline.

Another question is whether this additional on failure configuration (either a pipeline or specific options) should be at the processor level (each processor could take different actions) or unified for a pipeline. I tend to think that the latter is simpler to configure and covers most of the cases, with the additional benefit that it unifies what happens on failure for every processor in the pipeline. Having potentially a different behaviour for each processor might get confusing. Like the previous point, we should come up with good reasons why each processor should have a different behaviour, otherwise keep this at the pipeline level.

**on success**: in case everything works properly, do we need additional configuration for actions to be taken in case of success too? I don't think we do because in case of success the next processor in the chain will be executed, thus that next processor can be seen as the natural "on_success" processor.

I am marking this for discussion, it would be great to gather other thoughts on it.

**Comment by uboness** [comment]
wondering if it'd be sufficient to have the following logic:

- by default, a failure in a processor fails the document processing in the pipeline.
- it should be possible to set the following on every processor:

 ```json
 "on_error" : {
   "skip" : true | false,
   "mark" : true | false,
   "redirect" : <another pipeline name>
 }
 ```
- `skip` means the document continues to the next processor in the pipeline
- `mark` means the document is marked as failed (in a special meta data field)
- `redirect` will indicate that the document should be redirected to another pipeline. This will allow to have a "failure" pipeline that can potentially change the index name to a "failures" index and serve as a "catch all failures" strategy for multiple pipelines.
- `skip` and `redirect` are mutually exclusive.

**Comment by javanna** [comment]
Could we unify skip and mark, meaning that whenever we continue, we always have to mark otherwise the failure is lost? Would it make sense to continue without marking?

My other concern is still whether a redirect pipeline is needed or we can trim down what can be done to a few specific options instead. What you propose should be enough, not sure if too much? What are the usecases for a pipeline? Do you have examples in mind?

**Comment by uboness** [comment]
> Could we unify skip and mark, meaning that whenever we continue, we always have to mark otherwise the failure is lost? Would it make sense to continue without marking?

we could have `mark` set to `true` by default... I don't think it hurts (don't know all use cases... but what if you have a processor that is "a nice to have enrichment" and don't care much for errors there.

> My other concern is still whether a redirect pipeline is needed or we can trim down what can be done to a few specific options instead. What you propose should be enough, not sure if too much? What are the usecases for a pipeline? Do you have examples in mind?

tbh, I don't know what's enough. @talevy  can probably provide more input on that from the field. The rationale with a redirect pipeline is that it's as flexible as you want it and utilizes the same infra that we already have (pipelines).

**Comment by javanna** [comment]
We discussed this, here is the outcome: for phase 1 we will leave failure handling configuration at the pipeline level only. When a processor fails, we won't continue with the next processors in the pipeline in any case. Each pipeline will allow to configure a list of processors (rather than a whole pipeline specifying its id) to be executed in case of failure. By default (no failure processors configured) we will fail and return an error to the client (or fail that specific item in the bulk response); the document will not be indexed. Otherwise the specified on_failure processors will be executed, after which the document will be indexed, thus no error returned to the client.

Depends on #15036 so that the error is available as transient metadata field and can be referred to by e.g. the add processor so that it can be added to the document before it gets indexed.

**Comment by talevy** [comment]
After some discussion, a sort of consensus around an `on_failure` block for all processors to have was decided on, instead of a `continue_on_failure` flag described by this issue.

An example of this is as follows:

```
{
  description: "my pipeline with on_failure",
  processors: [
    {
      "grok" : {
          ...
          on_failure : [
            { "set" : { "grok.failed" : true } },
            { "meta": { "_index" : "failed-index" } }
          ]
      }
    },
    {
       "date" : {
         ...
       }
    }
  ]
}
```

an `on_failure` will also be added to a pipeline as a whole:

```
{
   "description" : "...",
   "processors" : [ ... ],
   "on_failure" : [ ... ]
}
```

processors without an `on_failure` parameter defined will throw an exception and exit the pipeline immediately. processors with `on_failure` defined will catch the exception and allow for further processors to run. Exceptions within the `on_failure` block will be treated the same as the top-level.

If a user wishes to handle a failure, and still exit the pipeline immediately, a `fail` processor will be introduced. This processor will do nothing but throw an exception so that the pipeline exits.

**Comment by talevy** [comment]
Since processors could exit with an exception at any point of their run time. The state of the IngestDocument can be incomplete. Since all existing processors are being moved to operate on one field at a time, this is not too much of an issue, but that is not a required rule. For example, a processor may be interested in added two fields, but fail while attempting to add the second field. This leaves the IngestDocument in a partially updated state before it was to be passed along to the `on_failure` block of processors. Is this a concern? If so, I am not sure what the best way of handling this would be. I believe it would require some "shadow" change-list that gets bundled up into a single transaction upon successful execution of a processor, otherwise there would not be a simple way to revert partial changes within a processor.

**Comment by javanna** [comment]
This has been implemented, every processor can be associated with a list of processors to be executed on failure. Same for a pipeline. The processor one has the precedence.
